### PR TITLE
Add client-side behaviour to investment projects filters

### PR DESCRIPTION
**Pull request by tyom**
Adds two new JS components:
- js-xhr: makes links with this class use XHR mechanism to update the page
- js-AutoSubmit - makes form with this class to automatically submit on change or enter

Adds new Nunjucks layout for XHR mechanism, partially updating the page instead of full reload.

Depends on #343.

### 💥 [Demo](https://datahub-beta2-pr-373.herokuapp.com/investment-projects) 💥

investment-filters

### CODE REVIEW

**Comment by tyom** [comment]
> It's big and might help to have a run through of the thinking behind how it works.

The client-side is actually not that big on its own. It's exploded due to the fact that its underlying branch has gone out of sync (fixups). It will shrink loads when I rebase.

File: src/server.js [hunk 1]
```diff
@@ -53,6 +53,11 @@ app.use(csrfToken())
 
 app.use(compression())
 
+app.use((req, res, next) => {
+  res.locals.isXhr = req.xhr
```
Comment (teneightfive) [inline src/server.js]: Why not add to the [locals middleware](https://github.com/uktrade/data-hub-frontend/blob/develop/src/middleware/locals.js)?

File: src/templates/_layouts/dit-base.njk [hunk 2]
```diff
@@ -113,7 +113,9 @@
             {% endblock %}
           {% endblock %}
           <div class="main-content__inner l-container">
-            {% block body_main_content %}{% endblock %}
+            <div id="xhr-outlet">
```
Comment (teneightfive) [inline src/templates/_layouts/dit-base.njk]: Does this element get replaced or does it replace the contents? Could add the id to `main-content__inner` if it doesn't get replaced?

File: assets/stylesheets/trumps/_utility.scss [hunk 3]
```diff
@@ -45,3 +45,30 @@
     float: right !important;
   }
 }
+
+.u-loading {
+
```
Comment (teneightfive) [inline assets/stylesheets/trumps/_utility.scss]: Remove line?

File: src/server.js (hunk 1 above)
Comment (tyom) [inline src/server.js]: True.

File: src/middleware/locals.js [hunk 4]
```diff
@@ -18,6 +18,7 @@ module.exports = function locals (req, res, next) {
     CURRENT_PATH: req.path,
     GOOGLE_TAG_MANAGER_KEY: config.googleTagManagerKey,
     BREADCRUMBS: breadcrumbItems,
+    isXhr: req.xhr,
```
Comment (teneightfive) [inline src/middleware/locals.js]: Const case?

File: assets/javascripts/lib/xhr.js [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+    }
+
+    return axios
+      .get(url, {
+        params,
+        headers: { 'X-Requested-With': 'XMLHttpRequest' },
+      })
+      .then(res => this.updateOutlet(res, params))
+      .catch(error => {
+        console.log(error)
```
Comment (teneightfive) [inline assets/javascripts/lib/xhr.js]: `console.error`? And what about if `console` isn't support?

**Pull request merged**